### ISSUE TITLE
Feature Request: Caching

### ISSUE THREAD

**Issue by eterry1388** [Shopify/erb-lint#158]
We have a large codebase of 1,277 erb files.  Running 12 linters on all of those files, it takes over 3 minutes. This is a bottleneck to our build system.  Ideally, we could run these in parallel or with caching (so only changed files are inspected).  This problem will only get worse as we enable more linters and add more erb files to our codebase.

Rubocop has a `cache` option explained here: https://github.com/rubocop-hq/rubocop/blob/master/manual/caching.md

We would love if such a feature existed for `erblint`.  Thank you!

**Comment by heqianw** [comment]
The suggested link was moved to this current location: https://github.com/rubocop-hq/rubocop/blob/master/docs/modules/ROOT/pages/usage/caching.adoc

**Comment by zachfeldman** [comment]
@eterry1388 an initial caching implementation is complete in https://github.com/Shopify/erb-lint/pull/268! If you, @heqianw , or anyone else following this thread want to give it a review and/or try out the branch, I'm definitely game for comments :+1:

**Comment by joshuapinter** [comment]
@zachfeldman Incredible! Thanks for putting in the work to get this done. I'll use your fork and see how it goes at https://github.com/cntral.

**Comment by joshuapinter** [comment]
> Question @zachfeldman: Where is the cache stored? In development this works out of the box but I'll want to use this in our Github Actions CI and will likely need to cache and restore it specifically in between setups.

Never mind, it's cached to `.erb-lint-cache` directory. I updated our CNTRAL CI to cache and restore this directory automatically to keep things fast. It'll take a few runs before all of our Runners' caches are warm.

I also added this to `.gitignore`. :+1:

**Comment by eterry1388** [comment]
Wanted to post one more test, this time with a ridiculous number of errors.

TL;DR
- Using cache when there are little to no errors greatly improves the speed of the run
- Using cache when there are a lot of errors greatly reduces the speed of the run, consumes a large amount of memory, and makes for very large cache files

## First run

```
time be bundle exec ../../personal/erb-lint/exe/erblint --lint-all --with-cache
Cache mode is on
Linting 461 files with 18 linters...

14206 error(s) were found in ERB files
bundle exec bundle exec ../../personal/erb-lint/exe/erblint --lint-all   29.23s user 1.24s system 84% cpu 36.071 total
```

## Second run

```
time be bundle exec ../../personal/erb-lint/exe/erblint --lint-all --with-cache
Cache mode is on
Linting 461 files with 18 linters...

14206 error(s) were found in ERB files
bundle exec bundle exec ../../personal/erb-lint/exe/erblint --lint-all   331.17s user 15.63s system 68% cpu 8:28.47 total
```

Yikes! With caching made it a LOT longer to run. I check out out the cache directory and some of the files were massive:

```
ls -alhS .erb-lint-cache
total 103232
-rw-r--r--    1 ericterry  staff   855K Sep 27 08:02 965f5b1d3d0c976a6ee368ac55d881185a8f306f
-rw-r--r--    1 ericterry  staff   609K Sep 27 08:01 0135f4d6d8b53ee8b87e1688489261ad09f98b19
-rw-r--r--    1 ericterry  staff   568K Sep 27 08:02 a2e1c15703c939c395640a39027aae4f83238cd0
-rw-r--r--    1 ericterry  staff   564K Sep 27 08:01 df23b118633784d2179883912dda76a75dc1a02c
...
```

This second run was using consistent high CPU the whole time (looking at `htop`) and the memory usage continued to grow slowly up to around 3.5 GB!

I wanted to run a third time, but this time without `--with-cache` to make sure it was an issue with this new feature as opposed to just a general feature with `erblint`

```
time be bundle exec ../../personal/erb-lint/exe/erblint --lint-all
Linting 461 files with 18 linters...

14206 error(s) were found in ERB files
bundle exec bundle exec ../../personal/erb-lint/exe/erblint --lint-all  27.94s user 0.87s system 89% cpu 32.256 total
```

It worked fine and memory usages stayed low.

**Comment by zachfeldman** [comment]
@joshuapinter @eterry1388 thanks so much for trying the cache and posting your results here! Super useful data to see how it works in other projects. As I mentioned in the pull request I unfortunately haven't had time to return to this as I'm only working on it in my spare time, but hope to soon.

**Comment by zachfeldman** [comment]
@eterry1388 I did some testing and I think I identified that the bottleneck is simply in how we store the cache files and restore them. If you take a look here https://github.com/Shopify/erb-lint/pull/268/files#diff-f47f3afa10e485d1d8ca903dd7f3a0da0b7dd4427efc465993d2f940cc5b82c6R40 you'll see I'm parsing the JSON structures in the cache files. I have to read each cache file, parse the JSON, then restore an `Offense` in some cases with the _full source_ of the original file. I'm sticking the full source from the already read source that the `cli.rb` reads earlier so that's probably not the bottleneck. But simple reading a bunch of cache files into memory then parsing their JSON is probably what's taking up all the RAM. Just running ERB Lint without the cache doesn't have to do any of that work, hence less memory usage.

I'm trying to reconsider how I store and retrieve the cache into a faster method than single files parsed as JSON with source mudged on top. Will report back if I think of something faster (maybe just using a single file? Maybe using YML? etc).

**Comment by zachfeldman** [comment]
@eterry1388 I reconsidered exactly *what* I was caching in this commit https://github.com/Shopify/erb-lint/pull/268 deciding to only cache the fields of an `Offense` necessary for the `Reporter` to print to the command line at the end of running `erb-lint`. I'm not sure if this'll cause other problems vs trying to restore the full `Offense` object but it seems to work well in my testing and reduce RAM usage a ton. No need to store huge JSON objects and restore them into memory!
https://github.com/Shopify/erb-lint/pull/268/commits/b331799a1a0543d29cdbf33a94a7a7da871a90c3

again, I'd like others to test this to verify it doesn't degrade any other functionality.

**Comment by eterry1388** [comment]
Ok. Here are my testing results.

## With everything passing

Your recent changes did not affect the speed at all, which is good. Also, when there are no errors, all the files in `.erb-lint-cache` are blank (well, technically `[]%`).

### First Run

```
time be bundle exec ../../personal/erb-lint/exe/erblint --lint-all --with-cache
Cache mode is on
Linting 461 files with 1 linters...

No errors were found in ERB files
bundle exec bundle exec ../../personal/erb-lint/exe/erblint --lint-all   22.09s user 1.03s system 84% cpu 27.492 total
```

### Second Run

```
time be bundle exec ../../personal/erb-lint/exe/erblint --lint-all --with-cache
Cache mode is on
Linting 461 files with 1 linters...

No errors were found in ERB files
bundle exec bundle exec ../../personal/erb-lint/exe/erblint --lint-all   1.29s user 0.81s system 46% cpu 4.545 total
```

## With a ton of errors

### First Run

```
time be bundle exec ../../personal/erb-lint/exe/erblint --lint-all --with-cache
Cache mode is on
Linting 457 files with 18 linters...

14113 error(s) were found in ERB files
bundle exec bundle exec ../../personal/erb-lint/exe/erblint --lint-all   27.77s user 1.06s system 87% cpu 33.036 total
```

### Second Run

```
time be bundle exec ../../personal/erb-lint/exe/erblint --lint-all --with-cache
Cache mode is on
Linting 457 files with 18 linters...

Exception occurred when processing: app/views/customers/index.html.erb
If this file cannot be processed by erb-lint, you can exclude it in your configuration file.
undefined method `to_sym' for nil:NilClass
/Users/ericterry/git/personal/erb-lint/lib/erb_lint/cached_offense.rb:35:in `from_json'
/Users/ericterry/git/personal/erb-lint/lib/erb_lint/cache.rb:22:in `block in get'
/Users/ericterry/git/personal/erb-lint/lib/erb_lint/cache.rb:21:in `map'
...[truncated backtrace]...

6095 error(s) were found in ERB files
bundle exec bundle exec ../../personal/erb-lint/exe/erblint --lint-all   2.10s user 1.39s system 47% cpu 7.325 total
```

There were a lot of exceptions. I'm just showing one of them above. However, they all seemed to be the same exception.  As you can see, the number of errors went down a lot. I think it is probably because of the exceptions and erb-lint not being able to process the file. However, it processed it just fine on the first run.

As far as the file sizes go, they have been greatly reduced:

```
ls -alhS .erb-lint-cache
total 5544
-rw-r--r--    1 ericterry  staff    28K Oct 11 19:51 44ee571f2159b1a7635a77d20c88497ccf29c051
-rw-r--r--    1 ericterry  staff    21K Oct 11 19:51 9ccf865c98c87342ec0560b33e884b00f531b38e
-rw-r--r--    1 ericterry  staff    20K Oct 11 19:51 2c6876d51828d2799979c7ffcaa2e766e4c8c3c0
-rw-r--r--    1 ericterry  staff    19K Oct 11 19:51 770a4950bf01262da1d646ea7fac2520d8acac96
-rw-r--r--    1 ericterry  staff    19K Oct 11 19:51 7e69231c130c783454ebdc3ca49df21e996e7c7e
...[truncated]...
```

## The fix

I got it to work without exceptions!

```
14113 error(s) were found in ERB files
bundle exec bundle exec ../../personal/erb-lint/exe/erblint --lint-all   1.55s user 0.93s system 46% cpu 5.275 total
```

Looks like you are trying to call `to_sym` in `from_json` method like this:

```ruby
parsed_json[:severity].to_sym
```

Just need to add a safe navigator operator:

```ruby
parsed_json[:severity]&.to_sym
```

This is because some of the `json` looks like this where `severity` is `nil`:

```ruby
{"message"=>"No space detected where there should be a single space.", "line_number"=>"22", "severity"=>nil}
```

**Comment by eterry1388** [comment]
Ok I did get all formatters to work with the `CachedOffense` class, but it about doubled the cached files:

```
ls -alhS .erb-lint-cache
total 7776
-rw-r--r--    1 ericterry  staff    48K Oct 11 21:21 44ee571f2159b1a7635a77d20c88497ccf29c051
-rw-r--r--    1 ericterry  staff    35K Oct 11 21:21 9ccf865c98c87342ec0560b33e884b00f531b38e
-rw-r--r--    1 ericterry  staff    33K Oct 11 21:21 2c6876d51828d2799979c7ffcaa2e766e4c8c3c0
-rw-r--r--    1 ericterry  staff    33K Oct 11 21:21 770a4950bf01262da1d646ea7fac2520d8acac96
-rw-r--r--    1 ericterry  staff    32K Oct 11 21:21 d9943624c9bb9cd12e5993e03d2f077310b8aa8f
...
```

I don't think this is is a big deal as it's still 20 times smaller than the original approach, and the time it took to run stayed the same (around 2 seconds). I'll put up my code changes soon.

**Comment by zachfeldman** [comment]
@eterry1388 pull request merged.

**Comment by etiennebarrie** [comment]
We have #282 ready for that. 🎉
